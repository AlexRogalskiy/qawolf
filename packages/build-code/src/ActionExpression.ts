export class ActionExpression {
  code() {
    return "";
  }

  updatableCode() {
    return this.code();
  }
}
