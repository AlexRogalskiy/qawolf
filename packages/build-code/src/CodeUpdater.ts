import { buildWorkflow } from "@qawolf/build-workflow";
import { ElementEvent, Step } from "@qawolf/types";
import { sortBy } from "lodash";
import { buildStepsCode } from "./buildStepsCode";

export const CREATE_CODE_SYMBOL = "// 🐺 CREATE CODE HERE";

type ConstructorOptions = {
  device?: string;
  isTest?: boolean;
  name: string;
  url: string;
};

// TODO we need to update selectors too?

export class CodeUpdater {
  private _events: ElementEvent[] = [];
  private _options: ConstructorOptions;

  private _commitedStepIndex: number = 0;
  public _steps: Step[] = [];

  constructor(options: ConstructorOptions) {
    this._options = options;
  }

  private _buildWorkflow() {
    const workflow = buildWorkflow({
      device: this._options.device,
      events: sortBy(this._events, e => e.time),
      name: this._options.name,
      url: this._options.url
    });
    return workflow;
  }

  public get numPendingSteps() {
    return this._steps.length - this._commitedStepIndex;
  }

  public prepareSteps(events: ElementEvent[]) {
    this._events.push(...events);

    const workflow = this._buildWorkflow();
    const steps = workflow.steps.filter(step => step.isFinal);

    const newSteps = steps.slice(this._steps.length);
    newSteps.forEach(step => {
      this._steps.push(step);

      // TODO push selectors here too...
    });
  }

  public createSteps(code: string): string | null {
    const newSteps = this._steps.slice(this._steps.length);
    if (newSteps.length < 1) return code;

    // TODO log error if create symbol is not found (debounced)
    // code.replace()

    const codeToInsert = buildStepsCode(newSteps);
    return code.replace(CREATE_CODE_SYMBOL, codeToInsert);
  }
}
