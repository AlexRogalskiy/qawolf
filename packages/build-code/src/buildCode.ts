import { Workflow } from "@qawolf/types";
import { readFileSync } from "fs-extra";
import { compile } from "handlebars";
import { camelCase } from "lodash";
import { resolve } from "path";
import { formatLaunch } from "./formatLaunch";
import { formatStep } from "./formatStep";

type BuildCodeOptions = {
  test?: boolean;
  workflow: Workflow;
};

const scriptTemplate = compile(
  readFileSync(resolve(__dirname, "../static/script.hbs"), "utf8")
);

const testTemplate = compile(
  readFileSync(resolve(__dirname, "../static/test.hbs"), "utf8")
);

export const buildCode = (options: BuildCodeOptions) => {
  const { workflow } = options;

  const template = options.test ? testTemplate : scriptTemplate;

  const code = template({
    launch: formatLaunch(workflow.url, workflow.device),
    name: camelCase(workflow.name),
    steps: workflow.steps.map((step, i) => {
      const previousStep = i > 0 ? workflow.steps[i - 1] : null;

      return formatStep(step, previousStep);
    }),
    url: workflow.url
  });

  return code;
};
