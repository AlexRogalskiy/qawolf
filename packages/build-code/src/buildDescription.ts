import { Step } from "@qawolf/types";
import { describeDoc } from "@qawolf/web";
import { isNil } from "lodash";

export const buildDescription = (step: Step): string => {
  if (step.action === "scroll") {
    return `scroll`;
  }

  const description = describeDoc(step.html);

  const target = step.html.node;

  // link/input
  const tagName = `${target.name === "a" ? " link" : ` ${target.name}` || ""}`;

  if (step.action === "click") {
    return `click${description}${tagName}`;
  }

  if (step.action === "select") {
    return `select${description}`;
  }

  if (step.action === "type") {
    if (step.replace && (step.value === "" || isNil(step.value))) {
      return `clear${description}${tagName}`;
    }

    const value = step.value as string;
    if (value && value.indexOf("↓Enter") === 0) {
      return `Enter`;
    }

    if (value && value.indexOf("↓Tab") === 0) {
      return `Tab`;
    }

    return `type into${description}${tagName}`;
  }

  throw new Error(`Invalid step action ${step.action}`);
};
