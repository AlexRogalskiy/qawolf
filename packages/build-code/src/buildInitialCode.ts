import { camelCase } from "lodash";
import { buildLaunch } from "./buildLaunch";

export type InitialCodeOptions = {
  device?: string;
  isTest?: boolean;
  name: string;
  patchHandle: string;
  url: string;
};

type TemplateOptions = {
  launch: string;
  name: string;
  patchHandle: string;
};

const buildInitialScript = ({ name, launch, patchHandle }: TemplateOptions) => {
  const code = `const { launch } = require("qawolf");
const selectors = require("../selectors/${name}");

const ${name} = async browser => {
  ${patchHandle}
};

exports.${name} = ${name};

if (require.main === module) {
  (async () => {
    const browser = ${launch}
    await ${name}(browser);
    await browser.close();
  })();
}`;

  return code;
};

const buildInitialTest = ({ name, launch, patchHandle }: TemplateOptions) => {
  const code = `const { launch } = require("qawolf");
const selectors = require("../selectors/${name}");

describe('${name}', () => {
  let browser;

  beforeAll(async () => {
    browser = ${launch}
  });

  afterAll(() => browser.close());
  ${patchHandle}
});`;

  return code;
};

export const buildInitialCode = (options: InitialCodeOptions) => {
  const templateOptions = {
    launch: buildLaunch(options.url, options.device),
    name: camelCase(options.name),
    patchHandle: options.patchHandle
  };

  if (options.isTest) {
    return buildInitialTest(templateOptions);
  }

  return buildInitialScript(templateOptions);
};
