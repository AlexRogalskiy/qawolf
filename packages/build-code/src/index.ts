export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
