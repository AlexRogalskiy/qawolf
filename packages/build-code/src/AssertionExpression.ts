import { ActionExpression } from "./ActionExpression";

type ConstructorOptions = {
  action: ActionExpression;
  description: string;
};

export class AssertionExpression {
  private _action: ActionExpression;
  private _description: string;

  constructor({ action, description }: ConstructorOptions) {
    this._action = action;
    this._description = description;
  }

  private _formatIt() {
    const description = this._description;
    if (!description.includes(`"`)) return `"can ${description}"`;
    if (!description.includes("'")) return `'can ${description}'`;
    return JSON.stringify("can " + description);
  }

  code() {
    const code = `\nit(${this._formatIt()}, async () => {\n  ${this._action.code()}\n  });\n`;
    return code;
  }

  updatableCode() {
    return this._action.updatableCode();
  }
}
