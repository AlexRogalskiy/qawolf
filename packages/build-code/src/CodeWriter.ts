import { ElementEvent } from "@qawolf/types";
import { pathExists, readFile, outputFile, outputJson } from "fs-extra";
import { bold, green } from "kleur";
import { debounce } from "lodash";
import { join, dirname } from "path";
import { buildInitialCode, InitialCodeOptions } from "./buildInitialCode";
import { CREATE_CODE_SYMBOL, CodeUpdater } from "./CodeUpdater";
import { stepToSelector } from "./stepToSelector";

export type CodeWriterOptions = Omit<InitialCodeOptions, "createCodeSymbol"> & {
  codePath: string;
};

export class CodeWriter {
  private _options: CodeWriterOptions;
  private _pollingIntervalId?: NodeJS.Timeout;
  private _selectorsPath: string;
  // public for tests
  public _updater: CodeUpdater;
  private _updating: boolean = false;

  protected constructor(options: CodeWriterOptions) {
    this._options = options;

    // XXX support updating selectors that are moved
    this._selectorsPath = join(
      dirname(this._options.codePath),
      "../selectors",
      `${this._options.name}.json`
    );

    this._updater = new CodeUpdater(options);
  }

  public static async start(options: CodeWriterOptions) {
    const writer = new CodeWriter(options);
    await writer._createInitialCode();
    return writer;
  }

  protected async _createInitialCode() {
    const codeExists = await pathExists(this._options.codePath);
    if (codeExists) return;

    const initialCode = buildInitialCode({
      ...this._options,
      createCodeSymbol: CREATE_CODE_SYMBOL
    });
    await outputFile(this._options.codePath, initialCode, "utf8");
  }

  // public for testing
  public async _loadUpdatableCode() {
    const code = await readFile(this._options.codePath, "utf8");

    if (!CodeUpdater.hasCreateSymbol(code)) {
      this._logMissingCreateSymbol();
      return;
    }

    return code;
  }

  protected _logMissingCreateSymbol = debounce(
    () => {
      console.log(
        bold().red("Cannot update code without this line:"),
        CREATE_CODE_SYMBOL
      );
    },
    10000,
    { leading: true }
  );

  // public for testing
  public async _updateCode() {
    if (this._updating || this._updater.getNumPendingSteps() < 1) return;

    const code = await this._loadUpdatableCode();
    if (!code) return;

    this._updating = true;

    const updatedCode = this._updater.updateCode(code);
    await Promise.all([
      outputFile(this._options.codePath, updatedCode, "utf8"),
      this._updateSelectors()
    ]);

    this._updating = false;
  }

  // public for testing
  public async _updateSelectors() {
    const selectors = this._updater._steps.map((step, index) => ({
      // inline index so it is easy to correlate with the test
      index,
      ...stepToSelector(step)
    }));

    await outputJson(this._selectorsPath, selectors, { spaces: " " });
  }

  public async discard() {
    this.stopUpdatePolling();

    // TODO restore to original, or delete if there was no original
  }

  public prepare(events: ElementEvent[]) {
    this._updater.prepareSteps(events);
  }

  public async save() {
    this.stopUpdatePolling();
    // TODO prepare w/ final options

    // TODO...
    // if (this.options.debug) {
    //   await this.saveJson("events", events);
    //   await this.saveJson("workflows", workflow);
    // }

    console.log(green("saved:"), `${this._options.codePath}`);
  }

  public startUpdatePolling() {
    this._pollingIntervalId = setInterval(async () => {
      await this._updateCode();
    }, 100);
  }

  public stopUpdatePolling() {
    if (!this._pollingIntervalId) return;

    clearInterval(this._pollingIntervalId);
    this._pollingIntervalId = undefined;
  }
}
