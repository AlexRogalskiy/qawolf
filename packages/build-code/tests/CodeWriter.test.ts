import { CodeWriter, CodeWriterOptions } from "../src/CodeWriter";
import { outputFile, pathExists, readFile } from "fs-extra";
import { buildInitialCode } from "../src/buildInitialCode";
import { CREATE_CODE_SYMBOL } from "../src/CodeUpdater";

jest.mock("fs-extra");

const mockedPathExists = pathExists as jest.Mock<Promise<boolean>>;
const mockedOutputFile = outputFile as jest.Mock<Promise<void>>;

const options: CodeWriterOptions = {
  codePath: "/path/to/mytest.test.js",
  isTest: true,
  name: "mytest",
  url: "http://localhost:3000"
};

describe("CodeWriter._createInitialCode", () => {
  describe("no existing code", () => {
    it("writes initial code", async () => {
      mockedPathExists.mockResolvedValue(false);

      const writer = await CodeWriter.start(options);
      expect(mockedOutputFile.mock.calls[0]).toEqual([
        options.codePath,
        buildInitialCode({
          ...options,
          createCodeSymbol: CREATE_CODE_SYMBOL
        }),
        "utf8"
      ]);
      writer.dispose();
    });
  });

  describe("existing code", () => {
    it("does not write initial code", async () => {
      mockedPathExists.mockResolvedValue(true);
      mockedOutputFile.mockClear();

      const writer = await CodeWriter.start(options);
      expect(mockedOutputFile.mock.calls.length).toEqual(0);
      writer.dispose();
    });

    it("inserts the create symbol for scripts", () => {
      // TODO
    });

    it("inserts the create symbol for tests", () => {
      // TODO
    });
  });
});

describe("CodeWriter._loadUpdatableCode", () => {
  it("logs when the create symbol is not found", async () => {
    const mockedReadFile = readFile as jest.Mock<
      Promise<string | number | Buffer>
    >;
    mockedReadFile.mockResolvedValue("no code");

    const writer = await CodeWriter.start(options);

    const consoleSpy = jest.spyOn(global.console, "log").mockImplementation();
    await writer._loadUpdatableCode();

    writer.dispose();

    expect(consoleSpy.mock.calls[0]).toEqual([
      "[1m[31mCannot update code without this line:[22m[39m",
      CREATE_CODE_SYMBOL
    ]);

    consoleSpy.mockRestore();
  });
});

describe("CodeWriter._updateCode", () => {
  it("updates code", () => {
    // TODO
  });

  it("updates selectors", () => {
    // TODO
  });
});

describe("CodeWriter.discard", () => {
  it("restores the existing code", () => {
    // TODO
  });

  it("deletes the file if there was not existing code", () => {
    // TODO
  });
});

describe("CodeWriter.save", () => {
  it("includes non-final steps", () => {
    // TODO
  });

  it("removes the CREATE_CODE_SYMBOL", () => {
    // TODO
  });
});
