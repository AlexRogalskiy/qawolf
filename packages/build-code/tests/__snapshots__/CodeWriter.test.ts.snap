// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeWriter._updateCode updates code and selectors 1`] = `
Array [
  "/path/selectors/mytest.json",
  Array [
    Object {
      "html": Object {
        "ancestors": Array [],
        "node": "<input/>",
      },
      "index": 0,
    },
  ],
  Object {
    "spaces": " ",
  },
]
`;
